File: original/JGDL/JGDLAnimation.java
/*
=======================================================
JGDL - Java Game Development Library
Implementation of the class JCGDLAnimation.
Copyright 2003, Nology Softwares. All rights reserved.
=======================================================
*/

package JGDL;

public class JGDLAnimation extends JGDLObject
{
	//!Ponteiro para o main da GDL.
	public JGDLMain pr_Main = null;
	//!Indica que a animação recomeça ao acabar se true.
	public boolean bRepeat;
	//!Indices dos quadros de animação na imagem.
	public JGDLList Frames = new JGDLList();
	//!Número de quadros por segundo na animação.
	public int iFramesPerSecond;
	//!Acumulator de tempo (usado para trocar quadros de animação).
	public int uiTimeAccum;
	//!Índice do quadro atual no vetor de frames.
	private int iCurrentFrame;

	//---------------------------------------------------------------------------- 
	// Name: JGDLAnimation()
	// Desc: rutor padrão
	// Pams: nenhum
	//---------------------------------------------------------------------------- 
	public JGDLAnimation()
	{
		bRepeat				= false;
		iCurrentFrame		= 0;
		iFramesPerSecond	= 1;
		pr_Main				= null;
	}
	
	//---------------------------------------------------------------------------- 
	// Name: Release()
	// Desc: finaliza a animação
	// Pams: nenhum
	//---------------------------------------------------------------------------- 
	public boolean Release()
	{
		Frames.clear();
		Frames = null;
		
		return true;
	}
	
	//---------------------------------------------------------------------------- 
	// Name: Ended()
	// Desc: retorna true se a animação acabou
	// Pams: nenhum
	//---------------------------------------------------------------------------- 
	public boolean Ended()
	{
		return (bRepeat)? false : iCurrentFrame >= Frames.size();
	}
	
	//---------------------------------------------------------------------------- 
	// Name: Update()
	// Desc: Atualiza a animação
	// Pams: nenhum
	//---------------------------------------------------------------------------- 
	public void Update()
	{
		if(null != pr_Main)
		{
			//pega o intervalo entre os quadros
			 int iFrameInterval = (iFramesPerSecond != 0) ? 1000/iFramesPerSecond : 1;
	
			//Atualiza o acumulador de tempo
			uiTimeAccum += pr_Main.TimeHandler.iFrameTime;
	
			//Se o acumulador eh maior que o intervalo entre quadros
			if(uiTimeAccum >= iFrameInterval)
			{
				//Calcula quantos quadros se passaram
				int iSum = (int)(uiTimeAccum / iFrameInterval);
	
				//Soma os quadros ao quadro corrente
				if(bRepeat)
				{
	        		iCurrentFrame += iSum;
					iCurrentFrame %= Frames.size();
				}
				else
				{
					iCurrentFrame += (iCurrentFrame < (char)Frames.size())? iSum : 0;
				}
				//reduz o acumulador ao resto da sua divisão pelo intervalo
				uiTimeAccum %= iFrameInterval;
			}
		}
	}
	
	//---------------------------------------------------------------------------- 
	// Name: GetCurrentFrame()
	// Desc: Retorna o quadro corrente da animação
	// Pams: nenhum
	//---------------------------------------------------------------------------- 
	public int GetCurrentFrame()
	{
		Integer iRet = (Integer) (((iCurrentFrame >= 0) && (iCurrentFrame < (int)Frames.size())) ? Frames.get(iCurrentFrame) : Frames.get(Frames.size()-1));
		return iRet.intValue();
	}
	
	//---------------------------------------------------------------------------- 
	// Name: Reset()
	// Desc: Reseta a animação
	// Pams: nenhum
	//---------------------------------------------------------------------------- 
	public void Reset()
	{
		uiTimeAccum		= 0;
		iCurrentFrame = 0;
	}
	
	//---------------------------------------------------------------------------- 
	// Name: GetClone()
	// Desc: retorna um clone da animação
	// Pams: 
	//---------------------------------------------------------------------------- 
	public JGDLAnimation GetClone()
	{
		JGDLAnimation p_Clone	= new JGDLAnimation();
		if(null != p_Clone)
		{
			p_Clone.bRepeat					= bRepeat;
			p_Clone.iCurrentFrame		= iCurrentFrame;
			p_Clone.iFramesPerSecond = iFramesPerSecond;
			p_Clone.pr_Main					= pr_Main;
			p_Clone.uiTimeAccum			= uiTimeAccum;
	
			int iSize = Frames.size();
			for(int i = 0; i < iSize; i++)
			{
				p_Clone.Frames.add(Frames.get(i));
			}
			return p_Clone;
		}
		return null;	
	}
}

File: migration/js/jgdl/jgdl-animation.js
function JGDLAnimation() {
  this.pr_Main = null;
  this.bRepeat = false;
  this.Frames = [];
  this.iFramesPerSecond = 1;
  this.uiTimeAccum = 0;
  this._iCurrentFrame = 0;
}

JGDLAnimation.prototype.Update = function () {
  if (!this.pr_Main || !this.Frames.length) return;
  var interval = this.iFramesPerSecond ? 1000 / this.iFramesPerSecond : 1000;
  this.uiTimeAccum += this.pr_Main.TimeHandler.iFrameTime;
  if (this.uiTimeAccum >= interval) {
    var n = Math.floor(this.uiTimeAccum / interval);
    this.uiTimeAccum %= interval;
    if (this.bRepeat) {
      this._iCurrentFrame = (this._iCurrentFrame + n) % this.Frames.length;
    } else {
      this._iCurrentFrame = Math.min(this._iCurrentFrame + n, this.Frames.length);
    }
  }
};

JGDLAnimation.prototype.GetCurrentFrame = function () {
  if (!this.Frames.length) return 0;
  if (this._iCurrentFrame >= this.Frames.length) return this.Frames[this.Frames.length - 1];
  var f = this.Frames[this._iCurrentFrame];
  return typeof f === 'number' ? f : (f == null || f < 0 ? 0 : f);
};

JGDLAnimation.prototype.Ended = function () {
  if (this.bRepeat) return false;
  return this._iCurrentFrame >= this.Frames.length;
};

JGDLAnimation.prototype.Reset = function () {
  this.uiTimeAccum = 0;
  this._iCurrentFrame = 0;
};
